ikarus/finiteelements/mechanics/materials/hyperelastic/interface.hh

```diff
@@ -0,0 +1,330 @@
+      auto [lambdas, N]   = principalStretches(realCVec);
+
+      e.val      = dev_.storedEnergy(lambdas);
+      e.grad.val = (transformDeviatoricStresses(dev_.stresses(lambdas), N).transpose() / 2 * dualC).trace();
+      e.val.grad = e.grad.val;
+
+      const auto Cmoduli = transformDeviatoricTangentModuli(dev_.tangentModuli(lambdas), N);
+      Eigen::array<Eigen::IndexPair<Eigen::Index>, 2> double_contraction  = {Eigen::IndexPair<Eigen::Index>(2, 0),
+                                                                             Eigen::IndexPair<Eigen::Index>(3, 1)};
+      Eigen::array<Eigen::IndexPair<Eigen::Index>, 2> double_contraction2 = {Eigen::IndexPair<Eigen::Index>(0, 0),
+                                                                             Eigen::IndexPair<Eigen::Index>(1, 1)};
+      const auto tCdual                  = tensorView(dualC, std::array<Eigen::Index, 2>({3, 3}));
+      const auto tCdualT                 = tensorView(dualC2, std::array<Eigen::Index, 2>({3, 3}));
+      const auto prod                    = Cmoduli.contract(tCdual, double_contraction);
+      const Eigen::Tensor<double, 0> res = tCdualT.contract(prod, double_contraction2);
+      e.grad.grad                        = res(0) / 4.0; // extracting value of zero order tensor
+
+      return e;
+    } else
+      static_assert(Dune::AlwaysFalse<Derived>::value, "No fitting ScalarType.");
+  }
+
+  /**
+   * \brief A helper function to compute the deviatoric part of the stresses.
+   *
+   * \details While using AutoDiff, if the eigenvalues (principal stretches) of C are degenerated, the derivative can
+   * have certain singularities. This is circumvented here by explicitly updating the derivatives.
+   *
+   */
+  template <typename Derived, typename ST>
+  requires(std::same_as<typename Derived::Scalar, ST>)
+  auto deviatoricStress(const Eigen::MatrixBase<Derived>& C, const Eigen::Vector<ST, dim>& lambdasST,
+                        Eigen::Matrix<ST, dim, dim> NST) const {
+    if constexpr (not Concepts::AutodiffScalar<ST>) {
+      return transformDeviatoricStresses(dev_.stresses(lambdasST), NST);
+    } else if constexpr (std::is_same_v<ST, autodiff::dual>) {
+      constexpr int nVoigtIndices = 6;
+      Eigen::Vector<autodiff::dual, nVoigtIndices> g;
+      auto Cvec           = toVoigt(C.derived());
+      const auto realCVec = derivative<0>(Cvec);
+      auto realC          = fromVoigt(realCVec);
+      auto dualC          = fromVoigt(Cvec.unaryExpr([](const auto& v) { return v.grad; }).eval());
+      auto [lambdas, N]   = principalStretches(realC);
+
+      const auto stresses = toVoigt(transformDeviatoricStresses(dev_.stresses(lambdas), N));
+      const auto Cmoduli  = toVoigt(transformDeviatoricTangentModuli(dev_.tangentModuli(lambdas), N));
+      Eigen::Vector<double, nVoigtIndices> stressDirectionalDerivatrive = Cmoduli * toVoigt(dualC);
```

Comment:
const

Reply:
also above

Reply:
This was not marked with `const` as I modify it later.
```cpp
      Eigen::Vector<double, nVoigtIndices> stressDirectionalDerivatrive = Cmoduli * toVoigt(dualC);
      stressDirectionalDerivatrive.topRows<3>() /= 2.0;
```
I will add `const` at other places applicable